### PR TITLE
Improve Module Source

### PR DESCRIPTION
**Pull request by adamretter**
This cleans up and improves the Source classes which are responsible for loading XQuery Modules (amongst other things).

It also fixes at least one bug when loading XQuery Library Modules from the filesystem.

### CODE REVIEW

File: exist-core/src/main/java/org/exist/source/URLSource.java [hunk 1]
```diff
@@ -43,144 +42,141 @@
...
+        super(hashKey(url.toString()));
         this.url = url;
     }
-    
+
+    @Override
+    public String path() {
+        final String protocol = url.getProtocol();
+        final String host = url.getHost();
+        if (protocol.equals("file") && (host == null || host.length() == 0 || "localhost".equals(host) || "127.0.0.1".equals(host))) {
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/source/URLSource.java:68]: better? `"file".equals(protocol)`

File: exist-core/src/main/java/org/exist/source/URLSource.java [hunk 2]
```diff
@@ -43,144 +42,141 @@
...
+            return url.getFile();
+        }
+        return url.toExternalForm();
+    }
+
+    @Override
+    public String type() {
+        final String protocol = url.getProtocol();
+        final String host = url.getHost();
+        if (protocol.equals("file") && (host == null || host.length() == 0 || "localhost".equals(host) || "127.0.0.1".equals(host))) {
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/source/URLSource.java:78]: better? `"file".equals(protocol)`

File: exist-core/src/main/java/org/exist/xquery/ModuleContext.java [hunk 3]
```diff
@@ -129,7 +129,7 @@ private void setParentContext(final XQueryContext parentContext) {
                     } else if (".".equals(parentContext.moduleLoadPath)) {
                         if (!".".equals(dir)) {
                             if (dir.startsWith("/")) {
-                                setModuleLoadPath("." + dir);
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/ModuleContext.java:132]: LGTM, interesting old version

File: exist-core/src/main/java/org/exist/xquery/functions/util/Eval.java [hunk 4]
```diff
@@ -283,13 +283,11 @@ private Sequence doEval(final XQueryContext evalContext, final Sequence contextS
...
         if (Type.subTypeOf(expr.getType(), Type.ANY_URI)) {
             String uri = null;
-            final Object key = querySource.getKey();
 
-            if (key instanceof XmldbURI) {
-                uri = XmldbURI.EMBEDDED_SERVER_URI.append(((XmldbURI) key).removeLastSegment()).toString();
-//          } else if (key instanceof URL) {
-//              TODO: uri = ((URL) key).getParent()
-            } else if (key instanceof String && querySource instanceof FileSource) {
+            if (querySource instanceof DBSource) {
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/functions/util/Eval.java:287]: DBsource can be remote too right?

**Review by dizzzz — APPROVED** [review]
LGTM

**Comment by dizzzz** [comment]
Please could you rebase?

File: exist-core/src/main/java/org/exist/xquery/functions/util/Eval.java (hunk 4 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/xquery/functions/util/Eval.java:287]: I am not quite sure what you are asking? But no... DBSource is always local. I have not changed the previous behaviour here.

File: exist-core/src/main/java/org/exist/source/URLSource.java (hunk 2 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/source/URLSource.java]: Done

File: exist-core/src/main/java/org/exist/source/URLSource.java (hunk 1 above)
Comment (adamretter) [reply exist-core/src/main/java/org/exist/source/URLSource.java]: Done

File: exist-core/src/main/java/org/exist/xquery/functions/util/Eval.java (hunk 4 above)
Comment (dizzzz) [reply exist-core/src/main/java/org/exist/xquery/functions/util/Eval.java:287]: ok thnx; just asking.

**Comment by dizzzz** [comment]
waiting for build.....

**Comment by dizzzz** [comment]
looks we have a problem on windows

image

**Review by dizzzz — CHANGES_REQUESTED** [review]
windows issues

**Comment by adamretter** [comment]
@dizzzz Thanks. The Windows issue should be fixed now.

**Review by dizzzz — APPROVED** [review]

**Comment by adamretter** [comment]
@dizzzz Thanks for reviewing and merging :-)

**Pull request merged**